src/cli/rustup_mode.rs

```diff
@@ -1159,9 +1166,71 @@ fn component_add(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
         let new_component = Component::new_with_target(component, false)
             .unwrap_or_else(|| Component::new(component.to_string(), target.clone(), true));
 
-        toolchain.add_component(new_component)?;
-    }
+        if m.is_present("allow-downgrade") {
+            let toolchain_date = toolchain
+                .get_manifest()
+                .ok()
+                .and_then(|m| m.map(|m| m.date))
+                .unwrap();
+            let downgrade_date = utc_from_manifest_date(&toolchain_date)
+                .unwrap_or_else(|| panic!("Malformed manifest date: {}", toolchain_date))
+                .pred();
+
+            let release_channel = match toolchain.name() {
+                s if s.starts_with("nightly") => "nightly",
+                s if s.starts_with("stable") => "stable",
+                s if s.starts_with("beta") => "beta",
+                _ => "nightly",
+            };
+            let nightly_date = format!(
+                "{}-{}",
+                &release_channel,
+                &downgrade_date.to_string()[0..10]
+            );
+            let temp_toolchain = cfg.get_toolchain(&nightly_date, false)?;
 
+            let status = if !temp_toolchain.is_custom() {
+                let components: Vec<_> = m
+                    .values_of("components")
+                    .map(|v| v.collect())
+                    .unwrap_or_else(Vec::new);
+                let targets: Vec<_> = m
+                    .values_of("targets")
+                    .map(|v| v.collect())
+                    .unwrap_or_else(Vec::new);
+                Some(temp_toolchain.install_from_dist(
+                    false,
+                    m.is_present("allow-downgrade"),
+                    &components,
+                    &targets,
+                )?)
+            } else if !temp_toolchain.exists() {
+                return Err(
+                    ErrorKind::InvalidToolchainName(temp_toolchain.name().to_string()).into(),
+                );
+            } else {
+                None
+            };
+
+            if let Some(status) = status.clone() {
+                println!();
+                common::show_channel_update(cfg, temp_toolchain.name(), Ok(status))?;
+            }
+
+            use rustup::UpdateStatus;
+            if let Some(UpdateStatus::Installed) = status {
+                temp_toolchain.make_default()?;
+                temp_toolchain.add_component(new_component)?;
+                toolchain.make_default()?;
+            } else if let Some(UpdateStatus::Unchanged) = status {
+                temp_toolchain.make_default()?;
+                temp_toolchain.add_component(new_component)?;
+                toolchain.make_default()?;
+            }
```

Comment:
Why are these pathways necessary?  If we passed through `install_from_dist` surely we will have added the listed components?  Also this odd changing of the default toolchain worries me, in case the inner `add_component()` calls fails, the default toolchain would have been unexpectedly altered to one which doesn't exist.  That feels wrong.